hnix-store-core/src/System/Nix/Internal/Hash.hs

```diff
@@ -0,0 +1,152 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module System.Nix.Internal.Hash where
+
+import qualified Crypto.Hash.MD5       as MD5
+import qualified Crypto.Hash.SHA1      as SHA1
+import qualified Crypto.Hash.SHA256    as SHA256
+import qualified Data.ByteString       as BS
+import qualified Data.ByteString.Char8 as BSC
+import           Data.Bits             (xor)
+import qualified Data.ByteString       as BS
+import qualified Data.ByteString.Lazy  as BSL
+import qualified Data.Hashable         as DataHashable
+import           Data.List             (foldl')
+import           Data.Proxy            (Proxy(Proxy))
+import qualified Data.Text             as T
+import qualified Data.Text.Encoding    as T
+import qualified Data.Vector           as V
+import           Data.Word             (Word8)
+
+-- | A tag for different hashing algorithms
+--   Also used as a type-level tag for hash digests
+--   (e.g. @Digest SHA256@ is the type for a sha256 hash)
+data HashAlgorithm
+  = MD5
+  | SHA1
+  | SHA256
+
+
+-- | Types with kind @HashAlgorithm@ may be a @HasDigest@ instance
+--   if they are able to hash bytestrings via the init/update/finalize
+--   API of cryptonite
+--
+--   Each instance defined here simply defers to one of the underlying
+--   monomorphic hashing libraries, such as `cryptohash-sha256`.
+class HasDigest (a :: HashAlgorithm) where
+
+  type AlgoCtx a :: *
+  hashLength        :: Proxy a -> Int
+
+  validateHashBytes :: BS.ByteString -> Maybe (Digest a)
+  validateHashBytes bs = if BS.length bs == hashLength (Proxy @a)
+                         then Just (Digest bs)
+                         else Nothing
+
+  initialize        :: AlgoCtx a
+  update            :: AlgoCtx a -> BS.ByteString -> AlgoCtx a
+  finalize          :: AlgoCtx a -> Digest a
+
+  toBase32          :: Digest a -> T.Text
```

Comment:
I don't think this is a class method really... Should just be a method on `Digest a`s

Reply:
I'm using the class signal different types going to base32 in different ways (where MD5 and SHA1 simply base32Bytes their hash, while a SHA256 hash needs to do truncation first).

The fact that different types have different `base32` behaviors in my code might be a mistake in my mental model. Like the way I don't have a `Truncated` newtype in the API. If we _did_ have `newtype Truncated = T BS.Bytestring`, then I could see defining a few monomorphic funuctions `base32FromMD5` `base32FromTruncatedSHA256`, `base32FromSHA1`. A polymorphic `base32` function over `Digest a` would be able to produce some strings that aren't used by nix - like `base32 (someHash :: Digest 'SHA256)` <- that would be an "invalid" (not used by nix) `Text`.

Reply:
Thesis (Page 95) talks about the `valid : Path -> Hash` mapping, which does use untruncated SHA256 hashes. So yep, have to rethink this :)